### PR TITLE
Support staking/gov params update proposal with validation and full test coverage

### PR DESCRIPTION
**Pull request by thismars**
# Overview
This PR adds support for govParamsUpdateProposal and stakingParamsUpdateProposal functionalities in firma-js. Core logic has been implemented and all tests have been successfully completed to prepare for the v0.3.2 release.

# Key Changes

### 1. Improved parameter validation using fast-deep-equal
- If any parameter is omitted, it is compared against default values using equal from fast-deep-equal.
- If the input is invalid, an error is thrown.
- Both success and failure test cases are included.

### 2. New utility function for parsing duration parameters
- Added FirmaUtil.parseDurationToString() to handle proper formatting for blockchain-compatible duration strings.
- Includes unit tests.

### 3. Custom interface for proposal parameters
- Defined dedicated interfaces for structured and type-safe proposal payloads.

### 4. Full test coverage with screenshot evidence
- Ran the full test suite, not limited to only the newly modified parts.
- Verified no regressions in v0.3.2 behavior.
- Screenshots of test results included.

# Test Results
- All unit and integration tests passed successfully
- Verified using `npm test` with full suite execution

## Summarize
- Total test cases: **322**
- Pass: **284**
- Skip: **33**
- Error: **5**

## Test output screenshots
- Wallet test
image

- Contract test
image

- Bank test
image

- Feegrant test
image

- Gas test
image

- IPFS & Nft test
image

- Staking test
image

- Distribution test
image

- Gov test
image

- Util & Chain & Slashing test
image

- Token test
image

- Authz test
image

- CosmWasm test
image

- Arbitary Sign & Mint test
image

- Cw20 test
image

- Cw721 test
image

- CwBridge test
image

- CwMarket test
image

- Failed test
image
  The test initially failed due to special requirements, but it was rerun separately after meeting the conditions and switching the wallet.

# Notes
- The implementation is based on Cosmos SDK v0.50 proposal structure.
- The structure is designed to be reusable for supporting future proposal types.

### CODE REVIEW

File: sdk/FirmaUtil.ts
Comment (haryunio) [inline sdk/FirmaUtil.ts:1]: these huge new utils should be covered with appropriate test cases.
recommend to add or edit test cases to ensure the result is reliable; especially edge case numbers.

File: sdk/FirmaGovService.ts [hunk 1]
```diff
@@ -449,7 +483,24 @@ export class FirmaGovService {
                 typeUrl: "/cosmos.gov.v1.MsgUpdateParams",
                 value: GovMsgUpdateParmas.encode(GovMsgUpdateParmas.fromPartial({
                     authority: FirmaGovService.GOV_AUTHORITY,
-                    params: params
+                    params: {
+                        minDeposit: params.min_deposit,
```
Comment (haryunio) [inline sdk/FirmaGovService.ts]: this logic is not to good. let's change into:
1. check if there is missing params
2. throw error
3. recommend users to use `firma.Gov.getParam();` and edit desired params
4. also should apply this logic to firma-station

**Comment by haryunio** [comment]
Re-run all test cases since we merged gov param changes.
Also add always-failing cases for proper fromPartial based filtering.

**Comment by thismars** [comment]
> Re-run all test cases since we merged gov param changes. Also add always-failing cases for proper fromPartial based filtering.

I'll include all the test results in the PR description.
The failure case has been added to 16.gox_tx.test.ts, and the test has already been completed.

**Review by haryunio — COMMENTED** [review]
Check these points to be fixed in the latest PR

File: sdk/FirmaUtil.ts
Comment (haryunio) [reply sdk/FirmaUtil.ts:1]: Reviewed changes from PR #11

File: sdk/FirmaGovService.ts (hunk 1 above)
Comment (thismars) [reply sdk/FirmaGovService.ts]: Reviewed changes from PRs #10 , #11 

1. used the fast-deep-equal module to compare the full set of parameters with:
```
if (!equal(requestedParams.params, fromPartialParams.params))
```

2. When a mismatch is detected, I throw an explicit error like this:

```
if (!equal(requestedParams.params, fromPartialParams.params)) {
   throw new Error("All governance parameters must be provided. Use getParamAsGovParams() to get current values and override only the parameters you want to change.");
}
```

3. Along with the error message above, I also provided example test cases using `getParamAsGovParams()` to clearly communicate the intended usage.

4. The updated logic has been applied in `firma-js`, and I plan to share it with the relevant stakeholders for implementation in `firma-station`.

**Review by haryunio — DISMISSED** [review]
LGTM

**Comment by thismars** [comment]
Please review pr

@koofree

File: sdk/firmachain/gov/GovQueryClient.ts [hunk 2]
```diff
@@ -9,65 +11,32 @@ export enum ProposalStatus {
...
-            amount: string;
-        }[],
-        max_deposit_period: string;
-    };
-    tally_params: {
-        quorum: string,
-        threshold: string,
-        veto_threshold: string,
-    };
+export interface GovParamType {
```
Comment (koofree) [inline sdk/firmachain/gov/GovQueryClient.ts:15]: I think this is a new Type Convention. Is it having any reason make a new Convention of the interface type name?

File: sdk/FirmaUtil.ts [hunk 3]
```diff
@@ -410,6 +411,171 @@ export class FirmaUtil {
...
 	}
+
+    /**
+     * Parses a duration string to a Duration object.
+     * Supports formats like "336h0m0s", "21d", "1000ms", "1s", "1m", "1h", "1ns", "1µs"/"1us"
+     * 
+     * @param durationStr - Duration string to parse (e.g., "336h0m0s", "21d")
+     * @returns Duration object with seconds and nanos fields
+     */
+    static parseDurationString(durationStr: string): { seconds: bigint; nanos: number } {
```
Comment (koofree) [inline sdk/FirmaUtil.ts:422]: Recommend using https://github.com/fullcalendar/temporal-polyfill
For the official JavaScript library, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal

File: sdk/FirmaUtil.ts [hunk 4]
```diff
@@ -410,6 +411,171 @@ export class FirmaUtil {
...
+        
+        // For commission rates, if it's effectively zero, return empty string
+        if (normalized === "") {
+            return "";
+        }
+
+        // Ensure the value is within valid commission rate range (0-1)
+        try {
+            const rate = parseFloat(normalized);
+            if (rate < 0) {
```
Comment (koofree) [inline sdk/FirmaUtil.ts]: this case can be combined like `commision rate value must be >=0 or <=1`

File: sdk/FirmaGovService.ts [hunk 5]
```diff
@@ -3,28 +3,33 @@ import {
...
+import {
+    MsgCancelProposal,
+    MsgSubmitProposal,
+    MsgUpdateParams as GovMsgUpdateParmas
+} from "@kintsugi-tech/cosmjs-types/cosmos/gov/v1/tx";
+import { MsgSoftwareUpgrade } from "@kintsugi-tech/cosmjs-types/cosmos/upgrade/v1beta1/tx";
+import { MsgCommunityPoolSpend } from "@kintsugi-tech/cosmjs-types/cosmos/distribution/v1beta1/tx";
+import { Proposal, Params as GovParams } from "@kintsugi-tech/cosmjs-types/cosmos/gov/v1/gov";
+import { Params as StakingParams } from "@kintsugi-tech/cosmjs-types/cosmos/staking/v1beta1/staking";
+// import { Params as GovParams } from "cosmjs-types/cosmos/gov/v1/gov";
```
Comment (koofree) [inline sdk/FirmaGovService.ts]: this comment is not required.

**Review by koofree — DISMISSED** [review]

**Comment by koofree** [comment]
I checked minor cases. And I approved properly.

File: sdk/firmachain/gov/GovQueryClient.ts (hunk 2 above)
Comment (thismars) [reply sdk/firmachain/gov/GovQueryClient.ts:15]: The variable formats used by the chain and the firma-js core are different, so a separate type was defined to bridge the gap.
For instance, values retrieved from the LCD API come in underscore style like max_validators, while the core uses camelCase style like maxValidators, making it hard to apply them directly.
To resolve this difference, a new type was introduced to unify the formats between both sides

File: sdk/FirmaUtil.ts (hunk 4 above)
Comment (haryunio) [reply sdk/FirmaUtil.ts]: [Applied](https://github.com/FirmaChain/firma-js/commit/32a42d39f1eb2a3e85af6c7e74c6b35a38a56a08) changes

File: sdk/FirmaGovService.ts (hunk 5 above)
Comment (haryunio) [reply sdk/FirmaGovService.ts]: [Applied](https://github.com/FirmaChain/firma-js/commit/1bde5797f78ea3c07952cd60af463cb654d8522b) changes

File: sdk/FirmaUtil.ts (hunk 3 above)
Comment (haryunio) [reply sdk/FirmaUtil.ts:422]: the received format is quite different from ISO-standard format and requires almost same processing after using the Temporl; therefore, this logic would be appropriate.

**Review by haryunio — APPROVED** [review]
LGTM: reviewed minor changes from the comment

**Comment by haryunio** [comment]
Will include further minor improvements in the next release; core logic was approved properly with more than two members, so will publish this.
